src/setup/node.rs

```diff
@@ -202,45 +198,43 @@ impl Node {
 
                         // Reply with headers for the initial block headers
                         let headers = blocks.iter().map(|block| block.header.clone()).collect();
-                        Message::Headers(Headers::new(headers))
-                            .write_to_stream(&mut stream)
+                        synthetic_node
+                            .send_direct_message(source, Message::Headers(Headers::new(headers)))
                             .await
                             .unwrap();
+
+                        source
                     }
-                    msg => panic!("Expected GetHeaders but got: {:?}", msg),
-                }
+
+                    (_, msg) => panic!("Expected GetHeaders but got: {:?}", msg),
+                };
 
                 // respond to GetData(inv) for the initial blocks
-                match filter.read_from_stream(&mut stream).await.unwrap() {
-                    Message::GetData(inv) => {
+                match synthetic_node.recv_message_timeout(TIMEOUT).await.unwrap() {
+                    (source, Message::GetData(inv)) => {
                         // The request must be for the initial blocks
                         let inv_hashes = blocks.iter().map(|block| block.inv_hash()).collect();
                         let expected = Inv::new(inv_hashes);
                         assert_eq!(inv, expected);
 
                         // Send the blocks
                         for block in blocks {
-                            Message::Block(Box::new(block))
-                                .write_to_stream(&mut stream)
+                            synthetic_node
+                                .send_direct_message(source, Message::Block(Box::new(block)))
                                 .await
                                 .unwrap();
                         }
                     }
-                    msg => panic!("Expected GetData but got: {:?}", msg),
-                }
 
-                Message::Ping(Nonce::default())
-                    .write_to_stream(&mut stream)
-                    .await
-                    .unwrap();
-                let filter =
-                    MessageFilter::with_all_auto_reply().with_ping_filter(Filter::Disabled);
-                match filter.read_from_stream(&mut stream).await.unwrap() {
-                    Message::Pong(_) => {}
-                    msg => panic!("Expected Pong but got: {:?}", msg),
+                    (_, msg) => panic!("Expected GetData but got: {:?}", msg),
                 }
+
+                synthetic_node.assert_ping_pong(source).await;
```

Comment:
Maybe add a comment about why we have this? Since it was a question you had as well :)

```
// Check that the node has received and processed all previous messages
```